### PR TITLE
Add pagination to GetAll response

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Retrieve a list of jobs for a service](https://tools.hmcts.net/jira/browse/RPE-146)

### Change description ###

- Remove redundant JobList wrapper
- Replace with spring framework pagination

**Does this PR introduce a breaking change?** (check one with "x")

```
[x] Yes
[ ] No
```

Only marking as yes because it is not compliant with current master endpoint

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java [hunk 1]
```diff
@@ -35,34 +39,71 @@
...
             .andExpect(status().isOk())
-            .andExpect(jsonPath("data[0].action").exists())
-            .andExpect(jsonPath("data[0].id").value(jobData.id))
-            .andExpect(jsonPath("data[0].name").value(jobData.name));
+            .andExpect(jsonPath("content[0].action").exists())
+            .andExpect(jsonPath("content[0].id").value(jobData.id))
+            .andExpect(jsonPath("content[0].name").value(jobData.name))
+            .andExpect(jsonPath("totalPages").value(1))
+            .andExpect(jsonPath("totalElements").value(1))
+            .andExpect(jsonPath("numberOfElements").value(1));
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java]: our api guide requires json field names to be snake_case

https://hmcts.github.io/restful-api-standards/#118

Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java]: 😱

File: src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java [hunk 2]
```diff
@@ -35,34 +39,71 @@
...
+            .andExpect(jsonPath("numberOfElements").value(0));
+        getResponse(0, 2)
+            .andExpect(jsonPath("totalPages").value(2))
+            .andExpect(jsonPath("numberOfElements").value(2));
+        getResponse(1, 2)
+            .andExpect(jsonPath("totalPages").value(2))
+            .andExpect(jsonPath("numberOfElements").value(1));
+        getResponse(1, 1)
+            .andExpect(jsonPath("totalPages").value(3))
+            .andExpect(jsonPath("numberOfElements").value(1));
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java]: I think it would be clearer if we'd set up how many jobs exist in 'given' step of this test

**Comment by rkondratowicz** [comment]
could you paste an example json that's being returned now?

File: src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java (hunk 2 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java]: thought about separating test cases. didn't appeal to me too much. but it's a no brainer to do so

**Comment by doncem** [comment]
sure thing.

```json
{
	"content": [{
		"id": "some-id",
		"name": "my-job-name",
		"action": {
			"url": "https://not-existing-service.gov.uk/do-stuff",
			"method": "POST",
			"headers": {
				"Authorization": "token-goes-here"
			},
			"body": null
		}
	}],
	"pageable": {
		"sort": {
			"sorted": false,
			"unsorted": true
		},
		"offset": 2,
		"pageNumber": 1,
		"pageSize": 2,
		"paged": true,
		"unpaged": false
	},
	"totalPages": 2,
	"totalElements": 3,
	"last": true,
	"size": 2,
	"number": 1,
	"sort": {
		"sorted": false,
		"unsorted": true
	},
	"first": false,
	"numberOfElements": 1
}
```

File: src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java (hunk 2 above)
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java]: I mean it can remain one test, but it never exlicitly states how many jobs there are to perform paging on

File: src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java (hunk 1 above)
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java]: I think there's a Jackson setting for that

File: src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java (hunk 2 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java]: Can put a comment or so. It is a single test suite with singular purpose - can scroll down to check the total. But I'll put it in the test "title"

File: src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java [hunk 3]
```diff
@@ -35,34 +39,71 @@
...
+            .andExpect(pageIsLast)
+            .andExpect(pageIsFirst);
         sendGet("/jobs?page=1&size=1")
             .andExpect(status().isOk())
-            .andExpect(content().json(emptyResponse));
+            .andExpect(contentIsArray)
+            .andExpect(contentIsEmpty)
+            .andExpect(pageIsLast)
+            .andExpect(pageIsFirst);
     }
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java]: we  could replace the 20+ lines above with:
```
Arrays.asList(
  "/jobs",
  "/jobs?page=1",
  "/jobs?size=1",
  "/jobs?page=1&size=1"
).foreach(url -> {
    sendGet(url)
      .andExpect(status().isOk())
      .andExpect(contentIsArray)
      .andExpect(contentIsEmpty)
      .andExpect(pageIsLast)
      .andExpect(pageIsFirst)
});
```

**Comment by doncem** [comment]
the response is now:
```json
{
	"content": [{
		"id": "some-id",
		"name": "my-job-name",
		"action": {
			"url": "https://not-existing-service.gov.uk/do-stuff",
			"method": "POST",
			"headers": {
				"Authorization": "token-goes-here"
			},
			"body": null
		}
	}],
	"pageable": {
		"sort": {
			"sorted": false,
			"unsorted": true
		},
		"offset": 2,
		"page_number": 1,
		"page_size": 2,
		"paged": true,
		"unpaged": false
	},
	"total_pages": 2,
	"last": true,
	"total_elements": 3,
	"size": 2,
	"number": 1,
	"sort": {
		"sorted": false,
		"unsorted": true
	},
	"first": false,
	"number_of_elements": 1
}
```

File: src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java [hunk 4]
```diff
@@ -34,35 +40,67 @@
     @MockBean private AuthService authService;
 
     @Test
-    public void should_return_no_data_when_empty_job_list_is_returned() throws Exception {
-        String emptyResponse = "{\"data\":[]}";
+    public void should_return_no_data_when_empty_job_list_is_returned() {
+        ResultMatcher contentIsArray = jsonPath("content").isArray();
+        ResultMatcher contentIsEmpty = jsonPath("content").isEmpty();
+        ResultMatcher pageIsLast = jsonPath("last").value(true);
+        ResultMatcher pageIsFirst = jsonPath("first").value(true);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java]: Those are now used once and can be inlined.

Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java]: yep 😇

File: src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java [hunk 5]
```diff
@@ -34,35 +39,62 @@
...
+        getResponse(0, 2)
+            .andExpect(jsonPath("total_pages").value(2))
+            .andExpect(jsonPath("number_of_elements").value(2));
+        getResponse(1, 2)
+            .andExpect(jsonPath("total_pages").value(2))
+            .andExpect(jsonPath("number_of_elements").value(1));
+        getResponse(1, 1)
+            .andExpect(jsonPath("total_pages").value(3))
+            .andExpect(jsonPath("number_of_elements").value(1));
     }
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java:99]: I think we should also test edge cases (like negative values, page index beyond number of pages available etc)

Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java:99]: beyond availability is there. i'll add the negative

File: src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java [hunk 6]
```diff
@@ -72,6 +105,29 @@ public void should_return_401_when_auth_token_is_invalid() throws Exception {
...
+
+        int minSize = JobsController.MIN_PAGE_SIZE - 1;
+
+        sendGet("/jobs?size=" + minSize)
+            .andExpect(status().isBadRequest())
+            .andExpect(jsonPath("error")
+                .value("must be greater than or equal to " + JobsController.MIN_PAGE_SIZE)
+            );
+
+        int maxSize = JobsController.MAX_PAGE_SIZE + 1;
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java:122]: It's bigger than 'max size'. How about `tooBigPageSize`? (and `tooSmallPageSize` above)

Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java:122]: semantics. can declare variously: `newMax`, `overMax`, `tooBig`, `maxBreach`.. in combination with `size` and `page` too. the point still clear among couple of lines.

let's say if some1 reads the code it's shorter and more correct:
> Let's declare that maximum size will be this and that then we are going to use it as a new limit in the request parameter...

Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/jobscheduler/controllers/jobscontroller/GetAllTest.java:122]: by sampling the code read, can see that probably best way, if necessary, is to expand to `newMaximumPageSize` with the highest count of characters

File: src/main/java/uk/gov/hmcts/reform/jobscheduler/controllers/JobsController.java [hunk 7]
```diff
@@ -31,6 +35,9 @@
 @RequestMapping(path = "/jobs")
 public class JobsController {
 
+    public static final int MIN_PAGE_SIZE = 1;
+    public static final int MAX_PAGE_SIZE = 100;
```
Comment (KarolJastrzebski) [inline src/main/java/uk/gov/hmcts/reform/jobscheduler/controllers/JobsController.java:39]: Minor refactoring idea: these constants could be moved to configuration.

**Review by KarolJastrzebski — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/jobscheduler/controllers/JobsController.java (hunk 7 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/jobscheduler/controllers/JobsController.java:39]: Could be. Still ideally would like to have hardcoded limit so the application user does not greedily overuse the values. The upper bound can be up to discussion/suggestion
